### ISSUE TITLE
Error with SingerCycleCollineation for Large Fields

### ISSUE THREAD

**Issue by Triagle** [gap-packages/FinInG#21]
Very similar issue to #14. The function `SingerCycleCollineation` will throw an error when the order of the field it's working with is too big. Here is a sample of the error you get,
```
gap> SingerCycleCollineation(2, 2^6);
Error, CVEC_INTREP_TO_CVEC: invalid object in list in
  CVEC_INTREP_TO_CVEC( List( l, function ( x )
        return CVEC_HandleScalar( c, x );
    end ), v ); at /usr/lib/gap/pkg/cvec-2.7.4/gap/cvec.gi:828 called from 
CVec( l[i], c ) at /usr/lib/gap/pkg/cvec-2.7.4/gap/cmat.gi:62 called from
NewMatrix( IsCMatRep, f, Length( m2[1] ), m2 
 ) at /usr/lib/gap/pkg/fining/lib/group.gi:430 called from
ProjElWithFrob( mat, IdentityMapping( gf ), gf 
 ) at /usr/lib/gap/pkg/fining/lib/group.gi:589 called from
CollineationOfProjectiveSpace( SingerCycleMat( n, q ), GF( q ) 
 ) at /usr/lib/gap/pkg/fining/lib/projectivespace.gi:2605 called from
<function "SingerCycleCollineation method">( <arguments> )
```
Took a bit of digging, but uncommenting [group.gi#429](https://github.com/gap-packages/FinInG/blob/568260e1410ce5ae199b97bc9257b1ad1164b35a/lib/group.gi#L429) fixes the problem by fixing the matrix representation before constructing the collineation. I can only assume that commenting out this line was done for a good reason, so my fix is probably not a good one.

As an aside, is there any reason why these functions aren't documented in the manual? They've been in the library for about 6 years, and besides this bug I've found they appear to work as expected.

**Comment by fingolfin** [comment]
I believe the line in question is commented out because it stems from the time before FinInG uses the `cvec` package internally. Roughly said, in the past, it was used to convert `A -> B`; now we convert `A -> C`; uncommenting the line causes us to convert `A -> B -> C`, which is less efficient. But it solves the problem, and I think that means there is a bug in the `cvec` conversion function. I'll take a look.

**Comment by fingolfin** [comment]
Dug some deeper. The problematic input executes this method:
```gap
InstallMethod( SingerCycleCollineation, [IsInt, IsInt],
	function(n, q)
	return CollineationOfProjectiveSpace(SingerCycleMat(n,q), GF(q));
end);
```
The problem down the line is that `SingerCycleMat(n,q)` has entries supposedly in GF(q), but not really:
```
gap> mat:=SingerCycleMat(2,2^6);
[ [ 0z, z0, 0z ], [ 0z, 0z, z0 ], [ 1+z2+z5+z7+z8+z10+z11+z13+z14+z16, 1+z3+z5+z6+z9+z13+z16+z17,
      1+z+z2+z3+z4+z6+z8+z9+z10+z11+z14+z15 ] ]
gap> mat[1,1] in GF(2^6);
true
gap> mat[3,3] in GF(2^6);
true
gap> mat[3,3]![1];
<a GF2 vector of length 18>
gap> mat[3,3]![2];
18
```

So these elements are in `GF(2^6)`, but they really were defined as elements of `GF(2^18)`, which just happen to live in the subfield `GF(2^6)`.

Next, `cvec` sees these elements, and as an optimization, it tries to directly use the low-level implementation of `IsCoeffsModConwayPolRep` and access the internal data I revealed above, expecting a list of *6* coefficients; but then it see 18 and "explodes".

So, let's see, whom to blame...

FinInG I blame the least, though perhaps `SingerCycleMat` could be implemented slightly differently to avoid working in the larger field when it is not necessary?  Let's look at what it does right now:
```gap
InstallMethod( SingerCycleMat, [ IsInt, IsInt],
	function(n, q)
	local basis, omega, mat;
	basis := Basis(AsVectorSpace(GF(q),GF(q^(n+1))));
	omega := Z(q^(n+1));
	# companion matrix
	mat := List(BasisVectors(basis), t -> Coefficients(basis, t*omega));
	return mat;
end);
```
So this could be changed to just this, which would also have the advantage of working for more pairs (n,q) (as it does not rely on expensive Conway Polynomials). The drawback would be that the return matrix would be different each time; I don't know if that's acceptable?
```gap
InstallMethod( SingerCycleMat, [ IsInt, IsInt ],
	function(n, q)
	local pol;
        pol := RandomPrimitivePolynomial(q, n+1);
        return TransposedMat(CompanionMat(pol));
end);
```

Back to the blame game: GAP was asked to produce coefficients in GF(2^6) and it did... kinda... but perhaps it should try harder to *really* produce them there (possibly using `FFECONWAY.TryToWriteInSmallerField`?)

Finally, `cvec`: It tries to be clever and directly work with elements in `IsCoeffsModConwayPolRep`, presumably because that's fast; but it's also dangerous, and easy to get wrong... Plus, the code does not even check for `IsCoeffsModConwayPolRep`, it just assumes that any FFE which is not `IsInternalRep` and not `IsZmodnZObj` always is `IsCoeffsModConwayPolRep` but that's not right.... Anyway, here is the code which I think is the root of the problems:
```gap
InstallGlobalFunction( CVEC_HandleScalar, function(cl,s)
  # cl is a cvecclass and s a scalar
  local v,d;
  if   IsInternalRep(s) then return s;
    # Note that this case also covers integers!
  elif IsZmodnZObj(s) then return s![1];
  fi;
  # Now we have to check, whether the field element is over the right field:
  d := cl![CVEC_IDX_fieldinfo]![CVEC_IDX_d];
  if s![2] <> d then
      s := FFECONWAY.WriteOverLargerField(s,d);
  fi;
  if IsGF2VectorRep(s![1]) then
    v := ShallowCopy(s![1]);
    PLAIN_GF2VEC(v);
    return v;
  elif Is8BitVectorRep(s![1]) then
    v := ShallowCopy(s![1]);
    PLAIN_VEC8BIT(v);
    return v;
  elif cl![CVEC_IDX_fieldinfo]![CVEC_IDX_p] < MAXSIZE_GF_INTERNAL then
    return s![1];
  else
    return List(s![1],x->x![1]);   # this unpacks ZmodnZObjs
  fi;
end );
```
As you can see, it even checks for the problem: but it only accounts for the case where the value `d` is too big; here it is too small. So, as a fix, we can use `FFECONWAY.TryToWriteInSmallerField` there. I'll release a new version of `cvec` with that fix ASAP.

**Comment by fingolfin** [comment]
Fixed in cvec 2.7.6 (just released) by https://github.com/gap-packages/cvec/pull/38
